File: src/io/ziheng/recursion/dynamicprogramming/FibonacciSequence.java
package io.ziheng.recursion.dynamicprogramming;

public class FibonacciSequence {
    public static void main(String[] args) {
        int n = 8;
        long startTime = System.currentTimeMillis();
        long result1 = FibonacciSequence.fib(n);
        long result2 = FibonacciSequence.fibWithTopToBottom(n);
        long result3 = FibonacciSequence.fibWithBottomToTop(n);
        long endTime = System.currentTimeMillis();
        assert result1 == result2 && result1 == result3;
        System.out.printf(
            "fib(%d): %d\n"
          + "fibTopToBottom(%d): %d\n"
          + "fibBottomToTop(%d): %d\n"
          + "cost: %f\n",
            n, result1,
            n, result2,
            n, result3,
            (endTime - startTime) / 1000.0
        );
    }
    /**
     * 斐波那契数列 - 自底向上动态规划。
     *
     * @param n
     * @return int
     */
    public static int fibWithBottomToTop(int n) {
        int[] dp = new int[n + 1];
        dp[0] = 0;
        dp[1] = 1;
        for (int i = 2; i <= n; i++) {
            dp[i] = dp[i - 1] + dp[i - 2];
        }
        return dp[n];
    }
    private static final int CAPACITY = 128;
    private static final int[] memo = new int[CAPACITY];
    static {
        for (int i = 0; i < CAPACITY; i++) {
            memo[i] = -1;
        }
    }
    /**
     * 斐波那契数列 - 自顶向下记忆化搜索。
     *
     * @param n
     * @return int
     */
    public static int fibWithTopToBottom(int n) {
        if (n == 0) {
            return 0;
        }
        if (n == 1) {
            return 1;
        }
        if (memo[n] != -1) {
            return memo[n];
        }
        memo[n] = fibWithTopToBottom(n - 1) + fibWithTopToBottom(n - 2);
        return memo[n];
    }
    /**
     * 斐波那契数列 - 标准递归版本。
     *
     * @param n
     * @return int
     */
    public static int fib(int n) {
        if (n == 0) {
            return 0;
        }
        if (n == 1) {
            return 1;
        }
        return fib(n - 1) + fib(n - 2);
    }
}
/* EOF */

File: src/cc/recursion/FibonacciSequence.cc
#include <stdio.h>
#include <time.h>
#include <vector>

using namespace std;

class FibonacciSequence {
private:
    int n;
    vector<int> memo;
public:
    int fib(int n);
    int fibTopToBottom(int n);
    int fibBottomToTop(int n);
    FibonacciSequence(int n);
    ~FibonacciSequence();
};

FibonacciSequence::FibonacciSequence(int n)
: n(n), memo(vector<int>(n + 1, -1)) {
    // ...
}
FibonacciSequence::~FibonacciSequence() {
    // ...
}
/**
 * 斐波那契数列 -> 暴力递归
 */
int FibonacciSequence::fib(int n) {
    if (n == 0) {
        return 0;
    }
    if (n == 1) {
        return 1;
    }
    return fib(n - 1) + fib(n - 2);
}
/**
 * 斐波那契数列 -> 记忆化搜索
 */
int FibonacciSequence::fibTopToBottom(int n) {
    if (n == 0) {
        return 0;
    }
    if (n == 1) {
        return 1;
    }
    if (memo[n] == -1) {
        memo[n] = fibTopToBottom(n - 1) + fibTopToBottom(n - 2);
    }
    return memo[n];
}
/**
 * 斐波那契数列 -> 动态递推
 */
int FibonacciSequence::fibBottomToTop(int n) {
    vector<int> dp = vector<int>(n + 1);
    dp[0] = 0;
    dp[1] = 1;
    for (int i = 2; i <= n; i++) {
        dp[i] = dp[i - 1] + dp[i - 2];
    }
    return dp[n];
}

int main(int argc, char const *argv[]) {
    int n = 42;
    FibonacciSequence fib = FibonacciSequence(n);

    time_t startTime = clock();
    int result1 = fib.fib(n);
    int result2 = fib.fibTopToBottom(n);
    int result3 = fib.fibBottomToTop(n);
    time_t endTime = clock();

    printf("fib(%d): %d, cost: %ld\n",
        n, result1, endTime - startTime
    );
    printf("fibTopToBottom(%d): %d, cost: %ld\n",
        n, result2, endTime - startTime
    );
    printf("fibBottomToTop(%d): %d, cost: %ld\n",
        n, result3, endTime - startTime
    );

    return 0;
}
/* EOF */